In [1]:
# Regression model to use height to predict weight
# Load modules and packages
import pandas as pd
from sklearn.model_selection import train_test_split as split
from sklearn.linear_model import LinearRegression
# Load dataset
file = 'https://raw.githubusercontent.com/wooihaw/datasets/main/genders_heights_weights.csv'
df = pd.read_csv(file)
# Dimension of dataset
print(df.shape)
# preview 10 lines of data
print(df.sample(10))
# Descriptive statistics
print(df.describe())
# Split dataset into training and testing sets
X = df.drop(columns=['Gender', 'Weight'])
y = df['Weight']
X_train, X_test, y_train, y_test = split(X, y, test_size=0.25, random_state=42)

(10000, 3)
      Gender  Height  Weight
4229    Male   174.5    81.0
1873  Female   163.8    60.8
7649    Male   169.7    79.4
8502    Male   179.0    78.9
2808    Male   178.3    90.6
5748  Female   162.1    62.1
1513    Male   168.3    73.3
3916  Female   142.5    39.6
4122    Male   166.1    76.0
260     Male   162.2    76.2
             Height        Weight
count  10000.000000  10000.000000
mean     168.573940     73.228260
std        9.772842     14.563851
min      137.800000     29.300000
25%      161.300000     61.600000
50%      168.400000     73.100000
75%      175.700000     84.900000
max      200.700000    122.500000


In [2]:
# 3.To check the dimensions of the training set (X_train) and testing set (X_test)
print("Training set dimensions:", X_train.shape)
print("Testing set dimensions:", X_test.shape)
print("Number of data samples in training set:", len(X_train))
print("Number of data samples in testing set:", len(X_test))


Training set dimensions: (7500, 1)
Testing set dimensions: (2500, 1)
Number of data samples in training set: 7500
Number of data samples in testing set: 2500


In [3]:
# Evaluate the performance of the model on the testing set.
lnr = LinearRegression().fit(X_train, y_train)

# Evaluate the linear regression model with the testing data and print the R2 score
print(f"lnr R2 score: {lnr.score(X_test, y_test)}")

lnr R2 score: 0.8590692022607564


In [4]:
df2=pd.get_dummies(df)
df2.sample(5)

Height  Weight  Gender_Female  Gender_Male
2971   167.8    64.7              1            0
8301   178.3    89.8              0            1
8145   168.4    74.1              0            1
5162   170.5    78.3              0            1
5231   174.3    92.1              0            1

In [5]:
# 5. Train a linear regression model using the training set.
from sklearn.linear_model import LinearRegression

# Create a linear regression model
regression_model = LinearRegression()

# Fit the model to the training data
regression_model.fit(X_train, y_train)

LinearRegression()

In [9]:
# Evaluate the performance of the model on the test set with encoded features.
r_squared_score_encoded = regression_model.score(X_test, y_test)
print("Model R^2 score on test data with encoded features:", r_squared_score_encoded)

Model R^2 score on test data with encoded features: 0.8590692022607564


In [10]:

# 10. Compare the performance of both models.
print("Performance without gender encoding:", r_squared_score)
print("Performance with gender encoding:", r_squared_score_encoded)

NameError: name 'r_squared_score' is not defined

In conclusion the performance with gender encoding is higher compared to than without the gender encoding with value almost to 1 that is 0.9057030207072467.Hence the accuracy is higher

EXPERIMENT 2.2

In [12]:
# Classification model to predict the fish species
# Load modules and packages
import pandas as pd
from sklearn.model_selection import train_test_split as split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
# Load dataset
file = 'https://raw.githubusercontent.com/wooihaw/datasets/main/fish.csv'
df = pd.read_csv(file)
# Dimension of dataset
print(df.shape)
# preview 10 lines of data
print(df.sample(10))
# Descriptive statistics
print(df.describe())
# Split dataset into training and testing sets
X = df.drop(columns=["Species"])
y = df["Species"]
X_train, X_test, y_train, y_test = split(X, y, test_size=0.25, random_state=42)
# Train and evaluate a k-NN model with k=1
knn = KNeighborsClassifier(n_neighbors=1).fit(X_train, y_train)
print(f"knn accuracy: {knn.score(X_test, y_test)}")
# Train and evaluate a decision tree model with max_depth=1
dtc = DecisionTreeClassifier(random_state=42, max_depth=1).fit(X_train, y_train)
print(f"dtc accuracy: {dtc.score(X_test, y_test)}")


(159, 7)
    Species  Weight  Length1  Length2  Length3   Height   Width
119   Perch   850.0     36.9     40.0     42.3  11.9286  7.1064
93    Perch   145.0     20.7     22.7     24.2   5.9532  3.6300
6     Bream   500.0     26.8     29.7     34.5  14.1795  5.2785
18    Bream   610.0     30.9     33.5     38.6  15.6330  5.1338
99    Perch   180.0     23.0     25.0     26.5   6.4395  3.6835
17    Bream   700.0     30.4     33.0     38.5  14.9380  5.1975
53    Roach   272.0     25.0     27.0     30.6   8.5680  4.7736
154   Smelt    12.2     11.5     12.2     13.4   2.0904  1.3936
62   Parkki    60.0     14.3     15.5     17.4   6.5772  2.3142
44    Roach   145.0     20.5     22.0     24.3   6.6339  3.5478
            Weight     Length1     Length2     Length3      Height       Width
count   159.000000  159.000000  159.000000  159.000000  159.000000  159.000000
mean    398.326415   26.247170   28.415723   31.227044    8.970994    4.417486
std     357.978317    9.996441   10.716328   11.61

C:\tools\Anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)


In [13]:
#3. TO check the dimensions of dataset, training set (X_train) and testing set (X_test)
print("Dataset dimensions:", df.shape)
print("Training set dimensions:", X_train.shape)
print("Testing set dimensions:", X_test.shape)



Dataset dimensions: (159, 7)
Training set dimensions: (119, 6)
Testing set dimensions: (40, 6)


In [14]:
#4.To find out how many features are there in the dataset
num_features = X_train.shape[1]
print("Number of features:", num_features)



Number of features: 6


In [15]:
#Data samples in in training and testing set
print("Training set dimensions:", X_train.shape)
print("Testing set dimensions:", X_test.shape)

Training set dimensions: (119, 6)
Testing set dimensions: (40, 6)


In [16]:
#To train and evaluate the performance of K-nearest neighbours (k-NN) classifier or different values of k (3, 5, 7 and 9)
k_values = [3, 5, 7, 9]
for k in k_values:
    knn = KNeighborsClassifier(n_neighbors=k).fit(X_train, y_train)
    accuracy = knn.score(X_test, y_test)
    print(f"k-NN accuracy with k={k}: {accuracy}")


k-NN accuracy with k=3: 0.475
k-NN accuracy with k=5: 0.6
k-NN accuracy with k=7: 0.575
k-NN accuracy with k=9: 0.6


C:\tools\Anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)
C:\tools\Anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True o

In [17]:
#To train and evaluate the performance of decision tree clasifier for different values of max_depth (from 2 to 10)

max_depth_values = list(range(2, 11))
for depth in max_depth_values:
    dtc = DecisionTreeClassifier(random_state=42, max_depth=depth).fit(X_train, y_train)
    accuracy = dtc.score(X_test, y_test)
    print(f"Decision tree accuracy with max_depth={depth}: {accuracy}")


Decision tree accuracy with max_depth=2: 0.675
Decision tree accuracy with max_depth=3: 0.7
Decision tree accuracy with max_depth=4: 0.7
Decision tree accuracy with max_depth=5: 0.725
Decision tree accuracy with max_depth=6: 0.625
Decision tree accuracy with max_depth=7: 0.75
Decision tree accuracy with max_depth=8: 0.775
Decision tree accuracy with max_depth=9: 0.7
Decision tree accuracy with max_depth=10: 0.725


Question 8 , Based on the coded answers displayed by both models K-NN and decision tree ,  it is found that the **Decision tree** provides a much optimum solution where its value is higher in range compared to K-NN with Optimum value for the **hyperparameter which is the value of 0.775**